### PR TITLE
Validate `ChunkWithProof` during deserialization

### PR DESCRIPTION
**Pull request by rafal-ch**

### CODE REVIEW

File: hashing/src/chunk.rs [hunk 1]
```diff
@@ -24,6 +59,20 @@ impl<const N: usize> ChunkWithProof<N> {
             .to_vec();
         Ok(ChunkWithProof { proof, chunk })
     }
+
+    pub fn is_valid(&self) -> bool {
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]: Probably should be private...

File: hashing/src/chunk.rs [hunk 2]
```diff
@@ -24,6 +59,20 @@ impl<const N: usize> ChunkWithProof<N> {
...
+        let chunk_hash = blake2b_hash(self.chunk());
+        self.proof
+            .merkle_proof()
+            .iter()
+            .next()
+            .map_or(false, |first_hash| chunk_hash == *first_hash)
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]:
```suggestion
        self.proof
            .merkle_proof()
            .first()
            .map_or(false, |first_hash| chunk_hash == *first_hash)
```

File: hashing/src/chunk.rs [hunk 3]
```diff
@@ -1,14 +1,49 @@
...
+        if candidate.is_valid() {
+            Ok(candidate)
+        } else {
+            Err(MerkleConstructionError::IncorrectChunkProof)
+        }
+    }
+}
+
 impl<const N: usize> ChunkWithProof<N> {
     pub const CHUNK_SIZE: usize = N;
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]:
```suggestion
```

We should just use `N` throughout this...

File: hashing/src/chunk.rs [hunk 4]
```diff
@@ -24,6 +59,20 @@ impl<const N: usize> ChunkWithProof<N> {
...
+            .iter()
+            .next()
+            .map_or(false, |first_hash| chunk_hash == *first_hash)
+    }
+
+    /// Get a reference to the chunk with proof's chunk.
+    pub fn chunk(&self) -> &[u8] {
+        self.chunk.as_slice()
+    }
 }
```
Comment (xcthulhu) [inline hashing/src/chunk.rs:92]:
```suggestion
const CHUNK_SIZE: usize = 1_048_576;   // 1 MiB

type CheckWithProof = SizedChunkWithProof<CHUNK_SIZE>;
```

File: hashing/src/chunk.rs [hunk 5]
```diff
@@ -1,14 +1,49 @@
...
+        };
+        if candidate.is_valid() {
+            Ok(candidate)
+        } else {
+            Err(MerkleConstructionError::IncorrectChunkProof)
+        }
+    }
+}
+
 impl<const N: usize> ChunkWithProof<N> {
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]:
```suggestion
impl<const N: usize> SizedChunkWithProof<N> {
```

Still not sure about a const-generic here... outside of testing in this file, a programmer should never construct a `ChunkWithProof` with `CHUNK_SIZE ≠ 1_048_576`.  How do we enforce that?

File: hashing/src/chunk.rs [hunk 6]
```diff
@@ -85,4 +133,44 @@ mod test {
         assert_eq!(proof_0.root_hash(), expected_root);
         assert_eq!(proof_1.root_hash(), expected_root);
     }
+
+    #[test]
+    fn validates_chunk_with_proofs() {
+        let large_data_size = (CHUNK_SIZE * 2) as usize;
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]: If `CHUNK_SIZE = 1`, is this really large?

File: hashing/src/primitives.rs [hunk 7]
```diff
@@ -106,6 +111,11 @@ where
     hash_pair(leaf_count_bytes, raw_root)
 }
 
+#[cfg_attr(
+    feature = "std",
+    derive(Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize,),
```
Comment (xcthulhu) [inline hashing/src/primitives.rs]: So `IndexedMerkleProof`s can be invalid too.  The validity conditions are:
  -  `index > count` 
  - `indexed_merkle_proof.merkle_proof.len() ≠ indexed_merkle_proof.compute_expected_proof_length()`
  
We should have a deserializer that checks this too...

File: hashing/src/chunk.rs [hunk 8]
```diff
@@ -85,4 +133,44 @@ mod test {
...
+
+    #[test]
+    fn validates_chunk_with_proof_after_deserialization() {
+        let large_data_size = (CHUNK_SIZE * 2) as usize;
+        let large_data = prepare_bytes(large_data_size);
+        let chunk_with_proof = ChunkWithProof::<CHUNK_SIZE>::new(large_data.as_slice(), 0).unwrap();
+
+        let json = serde_json::to_string(&chunk_with_proof).unwrap();
+        serde_json::from_str::<ChunkWithProof<CHUNK_SIZE>>(&json)
+            .expect("should deserialize correctly");
```
Comment (xcthulhu) [inline hashing/src/chunk.rs]:
```suggestion
        assert_eq!(
          chunk_with_proof,
          serde_json::from_str::<ChunkWithProof<CHUNK_SIZE>>(&json)
              .expect("should deserialize correctly"));
```

You could make a proptest...

File: hashing/src/chunk.rs (hunk 4 above)
Comment (rafal-ch) [reply hashing/src/chunk.rs:92]: Reverted to the initial solution which uses the `#[cfg(test)]` attribute.

File: hashing/src/chunk.rs (hunk 3 above)
Comment (rafal-ch) [reply hashing/src/chunk.rs]: Reverted to the initial solution which uses the #[cfg(test)] attribute.

File: hashing/src/chunk.rs (hunk 5 above)
Comment (rafal-ch) [reply hashing/src/chunk.rs]: That's a very good point, code is reverted to the initial solution which uses the #[cfg(test)] attribute.

File: hashing/src/chunk.rs (hunk 6 above)
Comment (rafal-ch) [reply hashing/src/chunk.rs]: Meaning: "large enough to not fit into a single chunk".
Renamed to something more meaningful, though.

File: hashing/src/chunk.rs (hunk 8 above)
Comment (rafal-ch) [reply hashing/src/chunk.rs]: Applied

Comment (rafal-ch) [reply hashing/src/chunk.rs]: I will turn most or all UT into proptests a bit later, after I'm sure that the UTs respects the expected behavior of the functions.

File: hashing/src/primitives.rs (hunk 7 above)
Comment (rafal-ch) [reply hashing/src/primitives.rs]: Added in 083549d98249bdcfbbe8337d712838fc7a6b0c26

**Comment by rafal-ch** [comment]
@xcthulhu all comments addressed. If there are more, I'd be happy to help :)

**Pull request merged**